tensorflow_privacy/privacy/membership_inference_attack/codelabs/codelab_privacy_risk_score.ipynb

```diff
@@ -0,0 +1,816 @@
+      "with 0.5 as the threshold on privacy risk score, the precision-recall pair is (0.5814285714285714, 0.814)\n",
+      "\n",
+      "For slice type: CLASS=5\n",
+      "with 0.8 as the threshold on privacy risk score, the precision-recall pair is (0.8387096774193549, 0.0052)\n",
+      "with 0.7 as the threshold on privacy risk score, the precision-recall pair is (0.7452229299363058, 0.0234)\n",
+      "with 0.6 as the threshold on privacy risk score, the precision-recall pair is (0.6262230919765166, 0.32)\n",
+      "with 0.5 as the threshold on privacy risk score, the precision-recall pair is (0.5760075737084122, 0.8518)\n",
+      "\n",
+      "For slice type: CLASS=6\n",
+      "with 0.6 as the threshold on privacy risk score, the precision-recall pair is (0.6118458884416331, 0.2128)\n",
+      "with 0.5 as the threshold on privacy risk score, the precision-recall pair is (0.5705128205128205, 0.712)\n",
+      "\n",
+      "For slice type: CLASS=7\n",
+      "with 0.6 as the threshold on privacy risk score, the precision-recall pair is (0.6164383561643836, 0.18)\n",
+      "with 0.5 as the threshold on privacy risk score, the precision-recall pair is (0.5659122874312748, 0.8852)\n",
+      "\n",
+      "For slice type: CLASS=8\n",
+      "with 0.5 as the threshold on privacy risk score, the precision-recall pair is (0.5507837390085383, 0.8644)\n",
+      "\n",
+      "For slice type: CLASS=9\n",
+      "with 0.6 as the threshold on privacy risk score, the precision-recall pair is (0.6332288401253918, 0.202)\n",
+      "with 0.5 as the threshold on privacy risk score, the precision-recall pair is (0.5841666666666666, 0.701)\n",
+      "\n",
+      "For slice type: CORRECTLY_CLASSIFIED=True\n",
+      "with 0.5 as the threshold on privacy risk score, the precision-recall pair is (0.5220848110408277, 0.6059067835717582)\n",
+      "\n",
+      "For slice type: CORRECTLY_CLASSIFIED=False\n",
+      "with 0.7 as the threshold on privacy risk score, the precision-recall pair is (0.7104532829471387, 0.3436936936936937)\n",
+      "with 0.6 as the threshold on privacy risk score, the precision-recall pair is (0.6816679555870855, 0.624024024024024)\n",
+      "with 0.5 as the threshold on privacy risk score, the precision-recall pair is (0.6396320935483625, 0.8274774774774775)\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "from tensorflow_privacy.privacy.membership_inference_attack.dataset_slicing import get_single_slice_specs\n",
+    "from tensorflow_privacy.privacy.membership_inference_attack.dataset_slicing import get_slice\n",
+    "slicing_spec = SlicingSpec(\n",
+    "                            entire_dataset = True,\n",
+    "                            by_class = True,\n",
+    "                            by_classification_correctness = True\n",
+    "                            )\n",
+    "input_slice_specs = get_single_slice_specs(slicing_spec, 10)\n",
+    "for single_slice_spec in input_slice_specs:\n",
```

Comment:
So is the individual level privacy risk score meant for slices or individual samples? I was naively expecting something like: out of all the samples in the training set, these are especially likely to be memorized. 

But this example makes me think that we're still doing this on groups of samples (like slices)

Reply:
We actually compute a privacy risk score for each individual sample: in data_structures.py, the SingleRiskScoreResult data class has input arguments of train_risk_scores and test_risk_scores, each element of them means a sample's privacy risk score.

I call the slice function mainly to follow what we do in detailed attacks, where the dataset is also divided into slices.
If you want, we can find time to chat more about this.